File: backend/src/main/java/org/kpet/petshop/Controllers/ReviewController.java
package org.kpet.petshop.Controllers;

import org.kpet.petshop.Models.Review;
import org.kpet.petshop.Models.User;
import org.kpet.petshop.Repositories.ReviewRepository;
import org.kpet.petshop.Repositories.UserRepository;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Product ratings/reviews. Reading is public (see SecurityConfig, same as
 * products); posting requires a logged-in account (any Client — no purchase
 * verification, this is a simple public review system, not a "verified
 * buyer" one). A user can leave at most one review per product; deleting is
 * self-or-admin, mirroring UserController's isSelfOrAdmin pattern.
 */
@RestController
@RequestMapping("/api/reviews")
public class ReviewController {

    private final ReviewRepository reviewRepository;
    private final UserRepository userRepository;

    public ReviewController(ReviewRepository reviewRepository, UserRepository userRepository) {
        this.reviewRepository = reviewRepository;
        this.userRepository = userRepository;
    }

    public record ReviewRequest(Long productId, int rating, String comment) {
    }

    public record RatingSummary(Long productId, double average, long count) {
    }

    @GetMapping(value = "/product/{productId}", produces = "application/json")
    public List<Review> getReviewsForProduct(@PathVariable Long productId) {
        return reviewRepository.findByProductIdOrderByCreatedAtDesc(productId);
    }

    // One aggregate (average + count) per product with at least one review —
    // lets the storefront grid show star ratings on every card without an API
    // call per product.
    @GetMapping(value = "/summary", produces = "application/json")
    public List<RatingSummary> getRatingSummary() {
        Map<Long, List<Review>> byProduct = reviewRepository.findAll().stream()
                .collect(Collectors.groupingBy(Review::getProductId));

        return byProduct.entrySet().stream()
                .map(entry -> {
                    List<Review> reviews = entry.getValue();
                    double average = reviews.stream().mapToInt(Review::getRating).average().orElse(0);
                    return new RatingSummary(entry.getKey(), average, reviews.size());
                })
                .toList();
    }

    @PostMapping(produces = "application/json")
    public ResponseEntity<?> createReview(@RequestBody ReviewRequest request, Authentication authentication) {
        if (request.rating() < 1 || request.rating() > 5) {
            return ResponseEntity.badRequest().body(Map.of("error", "La calificación debe ser entre 1 y 5."));
        }

        Long userId = Long.valueOf(authentication.getName());
        if (reviewRepository.existsByProductIdAndUserId(request.productId(), userId)) {
            return ResponseEntity.badRequest().body(Map.of("error", "Ya dejaste una reseña para este producto."));
        }

        User user = userRepository.findById(userId).orElse(null);
        String userName = user != null ? (user.getFirstName() + " " + user.getLastName()).trim() : "Cliente Kpet";

        Review review = new Review();
        review.setProductId(request.productId());
        review.setUserId(userId);
        review.setUserName(userName);
        review.setRating(request.rating());
        review.setComment(request.comment());

        return ResponseEntity.ok(reviewRepository.save(review));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteReview(@PathVariable Long id, Authentication authentication) {
        return reviewRepository.findById(id)
                .map(review -> {
                    boolean isAdmin = authentication.getAuthorities().stream()
                            .anyMatch(a -> a.getAuthority().equals("ROLE_ADMIN"));
                    boolean isOwner = authentication.getName().equals(String.valueOf(review.getUserId()));
                    if (!isAdmin && !isOwner) {
                        return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
                    }
                    reviewRepository.delete(review);
                    return ResponseEntity.ok().build();
                })
                .orElseGet(() -> ResponseEntity.notFound().build());
    }
}

File: backend-php/app/Http/Controllers/ReviewController.php
<?php

namespace App\Http\Controllers;

use App\Models\Review;
use App\Models\User;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\DB;

class ReviewController extends Controller
{
    public function byProduct($productId)
    {
        return response()->json(
            Review::where('productId', $productId)->orderByDesc('createdAt')->get()
        );
    }

    // Grouped in SQL (not loaded into PHP and grouped in memory) — same
    // output shape as the old RatingSummary: only products with >=1 review.
    public function summary()
    {
        $rows = DB::table('reviews')
            ->select('productId', DB::raw('AVG(rating) as average'), DB::raw('COUNT(*) as count'))
            ->groupBy('productId')
            ->get()
            ->map(fn ($r) => [
                'productId' => $r->productId,
                'average' => (float) $r->average,
                'count' => (int) $r->count,
            ]);

        return response()->json($rows);
    }

    public function store(Request $request)
    {
        $data = $request->all();
        $rating = (int) ($data['rating'] ?? 0);
        if ($rating < 1 || $rating > 5) {
            return response()->json(['error' => 'La calificación debe ser entre 1 y 5.'], 400);
        }

        $userId = $request->attributes->get('authUserId');
        $productId = $data['productId'] ?? null;

        $duplicate = Review::where('productId', $productId)->where('userId', $userId)->exists();
        if ($duplicate) {
            return response()->json(['error' => 'Ya dejaste una reseña para este producto.'], 400);
        }

        $user = User::find($userId);
        $userName = $user ? trim($user->firstName . ' ' . $user->lastName) : 'Cliente Kpet';

        $review = Review::create([
            'productId' => $productId,
            'userId' => $userId,
            'userName' => $userName,
            'rating' => $rating,
            'comment' => $data['comment'] ?? null,
            'createdAt' => now(),
        ]);

        return response()->json($review);
    }

    public function destroy(Request $request, $id)
    {
        $review = Review::find($id);
        if (!$review) {
            return response()->json(['error' => 'No encontrado'], 404);
        }

        $role = $request->attributes->get('authRole');
        $authUserId = $request->attributes->get('authUserId');
        $isSelfOrAdmin = strtolower((string) $role) === 'admin' || (string) $authUserId === (string) $review->userId;
        if (!$isSelfOrAdmin) {
            return response()->json(['error' => 'Acceso denegado'], 403);
        }

        $review->delete();
        return response()->noContent();
    }
}
